feat(auth): Email verification requirement for organizations

RyanSkonnord opened the pull request:
https://getsentry.atlassian.net/browse/ER-660

RyanSkonnord committed:
feat(auth): Email verification requirement for organizations

Incomplete, feature-flagged first pass.

RyanSkonnord committed:
Implement email verification compliance task

RyanSkonnord committed:
Revert typo

Co-authored-by: Josh Ferge <josh.ferge@gmail.com>

RyanSkonnord committed:
Refer to org flag during invite flow

RyanSkonnord committed:
Check in new feature flag

markstory commented on src/sentry/features/__init__.py:
```diff
@@ -107,6 +107,7 @@
 default_manager.add("organizations:related-events", OrganizationFeature)  # NOQA
 default_manager.add("organizations:relay", OrganizationFeature)  # NOQA
 default_manager.add("organizations:reprocessing-v2", OrganizationFeature)  # NOQA
+default_manager.add("organizations:required-email-verification", OrganizationFeature)  # NOQA
```
Should this be a flagr flag? It would need a trailing True for that.

RyanSkonnord replied:
👍

markstory commented on src/sentry/tasks/auth.py:
```diff
@@ -57,47 +59,70 @@ def email_unlink_notifications(org_id, actor_id, provider_key):
-            target_user=user,
-        )
+    def remove_non_compliant_members(self, org_id, actor_id, actor_key_id, ip_address):
+        org = Organization.objects.get(id=org_id)
```
```suggestion
        org = Organization.objects.get_from_cache(id=org_id)
```
You could use `get_from_cache()` to potentially save a DB hit.

RyanSkonnord replied:
To be clear, that would be a change to preexisting behavior on `remove_2fa_non_compliant_members`. But that should be safe.

markstory commented on src/sentry/tasks/auth.py:
```diff
@@ -57,47 +59,70 @@ def email_unlink_notifications(org_id, actor_id, provider_key):
+            logging_data = {"organization_id": org.id, "user_id": user.id, "member_id": member.id}
+
+            try:
+                member.remove_user()
```
Would this be using the new member flag statuses in the future?

RyanSkonnord replied:
Yes.

markstory commented on src/sentry/tasks/auth.py:
```diff
@@ -112,3 +137,46 @@ def _remove_2fa_non_compliant_member(member, org, actor=None, actor_key=None, ip
+    name="sentry.tasks.remove_2fa_non_compliant_members",
+    queue="auth",
+    default_retry_delay=60 * 5,
+    max_retries=5,
```
Are there any exceptions you'd like to catch and retry on? By default our instrumented_task decorator doesn't do any retries on its own.

markstory replied:
Celery can bind a [retry](https://docs.celeryproject.org/en/stable/userguide/tasks.html#retrying) method to a task that lets you capture exceptions and retry them manually. You can also use the [auto-retry](https://docs.celeryproject.org/en/stable/userguide/tasks.html#automatic-retry-for-known-exceptions) behavior.

markstory commented on static/app/data/forms/organizationSecurityAndPrivacyGroups.tsx:
```diff
@@ -29,6 +29,23 @@ export default [
         },
       },
+      {
+        /* TODO: Render only if "organizations:required-email-verification" feature flag is set */
```
You can use the `visible` hook on the field definition to do this.

RyanSkonnord replied:
👍

JoshFerge commented on src/sentry/api/invite_helper.py:
```diff
@@ -211,3 +209,33 @@ def accept_invite(self, user=None):
+            return False
+
+        user = self.request.user
+        primary_email_is_verified = (
```
could you use `User.get_verified_emails` here instead? and do we need to check the email returned is the primary? `UserEmail.is_primary`?

RyanSkonnord replied:
The filter clause `email=user.email` should have the effect of finding only the primary email (it uses the same logic as in `UserEmail.is_primary`).

Using `User.get_verified_emails` is an improvement. To keep it in a single query, I think we would want to do
```python
user.get_verified_emails().filter(email=user.email).exists()
```
rather than use `UserEmail.is_primary` (which is an application-level check).

Alternatively, we could prioritize clarity over efficiency with
```python
any(email.is_verified() for email in user.get_verified_emails())
```
but I hope that the `primary_email_is_verified` name makes it sufficiently clear.

JoshFerge commented on src/sentry/api/invite_helper.py:
```diff
@@ -211,3 +209,33 @@ def accept_invite(self, user=None):
+        return not primary_email_is_verified
+
+    def get_onboarding_steps(self) -> Dict[str, bool]:
+        return {
```
wonder if we could use a NamedTuple or small class to represent the onboarding steps.

RyanSkonnord replied:
I'm 👍 for a small data class in general. But I'm unsure what is the most Pythonic style when each instance gets its own function body. This lends itself to a proliferation of singleton classes, the same pattern as `OrganizationComplianceTask` in this PR. Which I don't think is bad exactly, but it's the kind of pattern that lends itself to, like, an abstract `enum` in Java and never seems as clean in Python. Alternatively, we could do something more similar to how it is now: have a series of stand-alone functions and pass each one to the constructor. Anyway, I'm open to suggestion.

In this particular case, I'm inclined to leave it as is under the [Rule of Three](https://en.wikipedia.org/wiki/Rule_of_three_%28computer_programming%29).

JoshFerge commented on src/sentry/models/organization.py:
```diff
@@ -127,6 +127,10 @@ class Organization(Model):
+                "Require and enforce email verification for all members.",
+            ),
         ),
         default=1,
```
not exactly related to your PR what does having `default=1` do on a BitField? is it automatically set to true then?

wedamija replied:
Bitfield is a bigint, so setting it to 1 just enabled the first bit

JoshFerge commented on src/sentry/tasks/auth.py:
```diff
@@ -112,3 +137,46 @@ def _remove_2fa_non_compliant_member(member, org, actor=None, actor_key=None, ip
+            # TODO: Need to send different email with additional info
+            #  that the user has been removed from the org and needs
+            #  to redo the invite flow?
+            user.send_confirm_email_singular(email)
```
wonder if we can do the changes to allow for restricted members prior to this going live. i think users are going to have a hard time with this flow unless we combine the verify email /invite into the same email

RyanSkonnord replied:
Prior to going live, sounds very preferable.

If our arm were twisted, we could at least include the verification link and invite link in the same email.

Now that I write that out, I realize that the *best* UX would be to click a single link that both verifies the email and rejoins you to the organization. It's equivalent from a security perspective, of course. Back-end support for that would be a bit of scope creep, though.

RyanSkonnord committed:
Merge branch 'master' into email-verification

RyanSkonnord committed:
Changes from code review

RyanSkonnord committed:
Merge branch 'master' into email-verification

RyanSkonnord committed:
Dummy out new flag

RyanSkonnord committed:
Get Organization from cache

Co-authored-by: Mark Story <mark@sentry.io>

JoshFerge reviewed (commented):
I'm kind of confused how we're supposed to use it in its current state -- it seems like we don't persist the option when the security form is saved and is based solely on the feature flag?

The option is also not in the organization serializer (src/sentry/api/serializers/models/organization.py) so even if the feature flag is flipped, it will always appear as though the option is turned off in security settings.

was there a strong reason for just not including the db level flag in this PR?

JoshFerge commented on src/sentry/tasks/auth.py:
```diff
@@ -57,47 +59,70 @@ def email_unlink_notifications(org_id, actor_id, provider_key):
+class TwoFactorComplianceTask(OrganizationComplianceTask):
+    log_label = "2FA"
+
+    def is_compliant(self, member: OrganizationMember) -> bool:
```
this function takes only member but above we pass it org and member which is causing it to error on my machine

JoshFerge replied:
is there any way we could add a test to this?

JoshFerge replied:
Edit: looks like there is a test and it is failing, disregard above

RyanSkonnord replied:
Fixed; thanks.

RyanSkonnord committed:
Merge branch 'master' into email-verification

RyanSkonnord committed:
Debug call to is_compliant

RyanSkonnord committed:
Revert "Dummy out new flag"

This reverts commit 68e90f4d7a37c54055b1128ae803928cfc748e6c.

RyanSkonnord committed:
Add requireEmailVerification to Organization model serializer

(Compare to src/sentry/api/endpoints/organization_details.py)

RyanSkonnord committed:
Add Organization.flags.require_email_verification

RyanSkonnord committed:
Merge branch 'master' into email-verification

RyanSkonnord commented:
> I'm kind of confused how we're supposed to use it in its current state -- it seems like we don't persist the option when the security form is saved and is based solely on the feature flag?
> 
> The option is also not in the organization serializer (src/sentry/api/serializers/models/organization.py) so even if the feature flag is flipped, it will always appear as though the option is turned off in security settings.
> 
> was there a strong reason for just not including the db level flag in this PR?

Long story short, it was a situationally motivated attempt to get the PR out the door without the added overhead of having a Django migration at all. The situation has changed (and the PR is still here) so I've put the flag back in. Sorry for the confusing state.

wedamija reviewed (approved):
Migration lgtm, just a no-op

RyanSkonnord committed:
Merge branch 'master' into email-verification

RyanSkonnord committed:
Add extra feature flag checks

These should be redundant to other checks but are extra bug-proofing.

RyanSkonnord committed:
Add check to keep org owner from kicking themselves out

RyanSkonnord committed:
Add email template for members kicked due to email verification

Combine the email verification link and the organization invite link
from two other email templates.

RyanSkonnord committed:
Merge branch 'master' into email-verification

RyanSkonnord committed:
Regenerate migration after merge

JoshFerge commented on src/sentry/tasks/auth.py:
```diff
@@ -112,3 +139,64 @@ def _remove_2fa_non_compliant_member(member, org, actor=None, actor_key=None, ip
+
+    def call_to_action(self, member: OrganizationMember):
+        user = member.user
+        if isinstance(user, User):
```
it seems that this check will result in an email not being sent initially to the user when they are kicked out, at least locally when i'm running it -- we remove the user from the member before calling `call_to_action`, so when it gets to here user is `None`. maybe `call_to_action` could optionally take in a user param?

RyanSkonnord replied:
Let me see if I can repro the `None` case. I believe I added the `isinstance` check to guard against an edge case where `user` is a `django.contrib.auth.models.User`, because `UserEmail.get_primary_email` will choke unless it's a `sentry.models.user.User`.

RyanSkonnord replied:
Oof, I think I see the problem -- I fixed this https://github.com/getsentry/sentry/pull/25972/commits/eab3d34ae93f7b32d045d9b83ed7dc4b13cff0e4 in the wrong direction. That's a heck of a silent failure; thanks @JoshFerge for the catch.

I'll tighten up the check and add a warning in the expected `django.contrib.auth.models.User` case.

RyanSkonnord committed:
Fix bug in call_to_action parameters

Improve check that caused a silent failure.

leedongwei reviewed (approved):
Some suggestions on the email copy, but you can just merge without them.

RyanSkonnord committed:
Merge branch 'master' into email-verification

RyanSkonnord committed:
Regenerate migration after merge

RyanSkonnord committed:
Tweaks to email copy

Co-authored-by: Danny Lee <dlee@sentry.io>

RyanSkonnord committed:
Merge branch 'master' into email-verification

RyanSkonnord committed:
Regenerate migration after merge (again)

RyanSkonnord committed:
Merge branch 'master' into email-verification
